### PR TITLE
chore(asm): store block callable into ASM context

### PR DESCRIPTION
**Pull request by juanjux**
# Motivation

Store the blocking function into the ASM request context so it can be further called by future blocking features, specially UserId blocking since it'll be implemented as functions (`isUserBlocked`, `blockUser`, `blockUserIfBlocked`) that the user will call at any moment. 

# Design

Each framework will store the block callable on the `_asm_request_context`. Tests have been added to make the call to it from Django or Flask views, but the user should not call them directly (the User ID blocking will provide functions for these).

## Checklist

- [X] Change(s) are motivated and described in the PR description.
- [X] Testing strategy is described if automated tests are not included in the PR.
- [X] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [X] Change is maintainable (easy to change, telemetry, documentation).
- [X] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [X] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [X] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: ddtrace/appsec/_asm_request_context.py [hunk 1]
```diff
@@ -59,6 +65,29 @@ def get_headers_case_sensitive():  # type: () -> bool
...
+    Calls or returns the stored block request callable, if set.
+    """
+    _callable = _DD_BLOCK_REQUEST_CALLABLE.get()
+    if _callable:
+        return _callable()
+
+    log.debug("Block request called but block callable not set by framework")
+
+
+def asm_request_context_set(remote_ip=None, headers=None, headers_case_sensitive=False, block_request_callable=None):
```
Comment (emmettbutler) [inline ddtrace/appsec/_asm_request_context.py:89]: What's going on here? This empty function looks like an unintentional change to me, but I may be missing something.

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/appsec/_asm_request_context.py (hunk 1 above)
Comment (emmettbutler) [inline ddtrace/appsec/_asm_request_context.py:89]: What's going on here? This empty function looks like an unintentional change to me, but I may be missing something.

**Review by emmettbutler — CHANGES_REQUESTED** [review]

File: ddtrace/appsec/_asm_request_context.py (hunk 1 above)
Comment (juanjux) [reply ddtrace/appsec/_asm_request_context.py]: Rebase fuckup and then going to do another urgent thing and forgot to set it as draft... thanks for the reminder.

**Comment by juanjux** [comment]
Did an additional change: the django block callable will also block the request (by raising a django `PermissionDenied` exception that will be checked on the `patch.py` to add out custom message. This way it works like the Flask callable and the user doesn't have to return the result in the view, which makes it easier to use when called on nested contexts of the user code.

**Review by christophe-papazian — APPROVED** [review]

**Review by christophe-papazian — DISMISSED** [review]

**Review by christophe-papazian — APPROVED** [review]

**Review by avara1986 — APPROVED** [review]

**Review by gnufede — APPROVED** [review]

File: ddtrace/appsec/_asm_request_context.py (hunk 1 above)
Comment (juanjux) [reply ddtrace/appsec/_asm_request_context.py]: @emmettbutler this is fixed, could you remove the request for change?

**Review by emmettbutler — APPROVED** [review]

**Pull request merged**